man/systemd.exec.xml

```diff
@@ -761,6 +761,28 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
         details.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>NUMAPolicy=</varname></term>
+
+        <listitem><para>Controls the NUMA memory policy of the executed processes. Takes a policy type, one of:
+        <option>default</option>, <option>prefer</option>, <option>bind</option>, <option>interleave</option> and
+        <option>local</option>. A list of NUMA nodes that should be used associated with the policy must be specified
+        in <varname>NUMAMask=</varname>. For more details on each policy please see,
+        <citerefentry><refentrytitle>set_mempolicy</refentrytitle><manvolnum>2</manvolnum></citerefentry>. For overall
+        overview of NUMA support in Linux see,
+        <citerefentry><refentrytitle>numa</refentrytitle><manvolnum>7</manvolnum></citerefentry>
+        </para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>NUMAMask=</varname></term>
+
+        <listitem><para>Controls the NUMA node list which will be applied alongside with selected NUMA policy.
+        Takes a list of NUMA nodes and has the same syntax as a list of CPUs for <varname>CPUAffinity=</varname>
+        option. Note that the list of NUMA nodes is not required for <option>default</option> and <option>local</option>
+        policies and for <option>prefer</option> policy we expect a single NUMA node.</para></listitem>
```

Comment:
I don't think it would be a good idea to pick policy automatically. Imagine I have a NUMAMask with just a single node and no specification for NUMAPolicy. I wouldn't know which policy to pick as admin might wanted strict policy (MPOL_BIND) however more relaxed policy (MPOL_PREFERRED) makes sense for mask with just one node too.

Reply:
Well, but the user can always pick the policy explicitly if they want control. I think there's nothing wrong with deriving some roughly sensible automatic defaults from settings if things are underspecified.

I mean, let's turn this around: unless I am mistaken using MPOL_DEFAULT but specifying a mask makes no sense whatsoever. Hence, if the user only specifies a mask, why not default to MPOL_PREFERRED or so (which appears to be a good default to me if a mask is specified)? we should make this all work with minimal effort if we can. i.e. make the common stuff easy and the complex stuff possible...

Reply:
If the policy is configured improperly (underspecified) I'd rather fail service startup instead of picking something and letting it run potentially misconfigured (in a sense that applied policy would violate user's expectations, even though he made a mistake and left things misconfigured he arguably had some expectations about what the outcome should be like).

I mean this stuff is going to be used only by handful of services. And will be always explicit user config. Distros will never ship unit files with NUMA stuff enabled by default. So I don't think that we need to be too user-friendly here. Moreover, I'd bet you that this user friendly defaults would be considered a bug by the crowd that is target audience for this feature. 

If you really insist and won't press the green button w/o this then I am going to do it but I don't agree I must say.

@keszybz any opinions ?

Reply:
I agree with @poettering here. `MPOL_PREFERRED` seems to be the correct choice almost always, so if user specifies the mask, we should default to that.